src/v/iceberg/manifest_entry_values.cc

```diff
@@ -0,0 +1,248 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#include "iceberg/manifest_entry_values.h"
+
+#include "bytes/iobuf_parser.h"
+#include "strings/string_switch.h"
+
+namespace iceberg {
+
+namespace {
+
+ss::sstring from_iobuf(iobuf b) {
+    auto size_bytes = b.size_bytes();
+    iobuf_parser parser{std::move(b)};
+    return parser.read_string(size_bytes);
+}
+
+template<typename PrimitiveV>
+auto get_required_primitive(value v) {
+    if (!holds_alternative<primitive_value>(v)) {
+        throw std::invalid_argument("Value is not a primitive");
+    }
+    auto& as_primitive = std::get<primitive_value>(v);
+    if (!holds_alternative<PrimitiveV>(as_primitive)) {
+        throw std::invalid_argument("Value is not the expected type");
+    }
+    auto& as_t = std::get<PrimitiveV>(as_primitive);
+    return std::move(as_t.val);
```

Comment:
pessimizing `std::move`?

Reply:
Left a note that this is used for iobufs. Could separate out the template for iobuf-like types if you prefer